core/src/main/java/org/elasticsearch/rest/RestRequest.java

```diff
@@ -254,8 +338,10 @@ public final XContentParser contentParser() throws IOException {
         BytesReference content = content();
         if (content.length() == 0) {
             throw new ElasticsearchParseException("Body required");
+        } else if (xContentType.get() == null) {
+            throw new IllegalStateException("no content-type has been set so we cannot create a parser");
```

Comment:
I do not think we can fail earlier right now and this will be removed in 6.0 when the backwards compat layer is removed